File: longestSubstringWithoutRepeatingChars.go
package main

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	maxLength := 0
	charIndex := make([]int, 128)

	start := 0
	for end := 0; end < n; end++ {
		start = max(start, charIndex[int(s[end])])
		maxLength = max(maxLength, end-start+1)
		charIndex[int(s[end])] = end + 1
	}

	return maxLength
}

File: longestSubstringWithoutRepeatingChars.js
function lengthOfLongestSubstring(s) {
    const n = s.length;
    let maxLength = 0;
    const charIndex = new Array(128).fill(0);

    let start = 0;
    for (let end = 0; end < n; end++) {
        start = Math.max(start, charIndex[s.charCodeAt(end)]);
        maxLength = Math.max(maxLength, end - start + 1);
        charIndex[s.charCodeAt(end)] = end + 1;
    }

    return maxLength;
}